synapse/lib/oauth.py

```diff
@@ -0,0 +1,398 @@
+
+logger = logging.getLogger(__name__)
+
+reqValidProvider = s_config.getJsValidator({
+    'type': 'object',
+    'properties': {
+        'iden': {'type': 'string', 'pattern': s_config.re_iden},
+        'name': {'type': 'string'},
+        'flow_type': {'type': 'string', 'default': 'authorization_code', 'enum': ['authorization_code']},
+        'auth_scheme': {'type': 'string', 'default': 'basic', 'enum': ['basic']},
+        'client_id': {'type': 'string'},
+        'client_secret': {'type': 'string'},
+        'scope': {'type': 'string'},
+        'auth_uri': {'type': 'string'},
+        'token_uri': {'type': 'string'},
+        'redirect_uri': {'type': 'string'},
+        'extensions': {
+            'type': 'object',
+            'properties': {
+                'pkce': {'type': 'boolean'},
+            },
+            'additionalProperties': False,
+        },
+        'extra_auth_params': {
+            'type': 'object',
+            'additionalProperties': {'type': 'string'},
+        },
+    },
+    'additionalProperties': False,
+    'required': ['iden', 'name', 'client_id', 'client_secret', 'scope', 'auth_uri', 'token_uri', 'redirect_uri'],
+})
+
+reqValidTokenResponse = s_config.getJsValidator({
+    'type': 'object',
+    'properties': {
+        'access_token': {'type': 'string'},
+        'expires_in': {'type': 'number', 'exclusiveMinimum': 0},
+    },
+    'additionalProperties': True,
+    'required': ['access_token', 'expires_in'],
+})
+
+class OAuthManager(s_nexus.Pusher):
+    '''
+    Organize and execute OAuth token refreshes.
+    '''
+
+    async def __anit__(self, cell, nexsroot=None):
+        await s_nexus.Pusher.__anit__(self, 'oauth', nexsroot=nexsroot)
+
+        self.cell = cell
+
+        slab = self.cell.slab
+        self.clients = s_lmdbslab.SlabDict(slab, db=slab.initdb('oauth:v2:clients'))     # key=<provideriden><useriden>
+        self.providers = s_lmdbslab.SlabDict(slab, db=slab.initdb('oauth:v2:providers')) # key=<provideriden>
+
+        self.refresh_window = 0.5
+
+        self.ssl = None
+        cadir = self.cell.conf.get('tls:ca:dir')
+        if cadir is not None:
+            self.ssl = s_common.getSslCtx(cadir)
+
+        self.timeout = aiohttp.ClientTimeout(total=10)
+
+        self.schedule_map = {}
+        self.schedule_heap = []
+        self.schedule_task = None
+        self.schedule_wake = asyncio.Event()
+
+        # For testing
+        self._schedule_empty = asyncio.Event()
+        self._schedule_item_ran = asyncio.Event()
+
+    async def initActive(self):
+        self._loadSchedule()
+
+    async def initPassive(self):
+        self._clearSchedule()
+
+    def _clearSchedule(self):
+        if self.schedule_task is not None:
+            self.schedule_task.cancel()
+            self.schedule_task = None
+
+        self.schedule_map.clear()
+        self.schedule_heap.clear()
+        self.schedule_wake.clear()
+
+    def _loadSchedule(self):
+        if not self.cell.isactive:
```

Comment:
when would we call this when we are not active?

Reply:
Leftover cruft from before I split out scheduling the individual items.  Removed.